README.md
# 快速开始

#### 安装

```bash
npm i web-storage-extend
```

#### 基本使用

```ts
// 引入
import LocalCache from 'web-storage-extend'

const localCache = new LocalCache()

// 有效期单位为秒
localCache.setCache('name', 'cake', 1)
localCache.getCache('name')
localCache.deleteCache('name')
localCache.clearCache()
```

#### 进阶使用

```js
const localCache = new LocalCache({
  type: 'localStorage',
  prefix: 'cake',
  isEncrypt: true
})
```



| 字段      | 解释       | 必传 | 类型    | value                          | 默认值       |
| --------- | ---------- | ---- | ------- | ------------------------------ | ------------ |
| type      | 存储类型   | 是   | string  | localStorage \| sessionStorage | localStorage |
| prefix    | 前缀       | 是   | string  | -                              | -            |
| expire    | 有效时间   | 否   | number  | -                              | 24 * 60 * 60 |
| isEncrypt | 加密、解密 | 是   | boolean | -                              | true         |



CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.5.2](https://github.com/coding327/webStorage/compare/v1.5.1...v1.5.2) (2023-06-23)

### [1.5.1](https://github.com/coding327/webStorage/compare/v1.5.0...v1.5.1) (2023-06-23)

## [1.5.0](https://github.com/coding327/webStorage/compare/v1.4.0...v1.5.0) (2023-06-23)


### Features

* 🚀 调整文档 ([1cd99cb](https://github.com/coding327/webStorage/commit/1cd99cbc74136c0487a35a379c3c09ccf30e9418))

## [1.4.0](https://github.com/coding327/webStorage/compare/v1.3.0...v1.4.0) (2023-06-23)


### Features

* 🚀 文档调整 ([1baef80](https://github.com/coding327/webStorage/commit/1baef80d70ea8abe574f0386b548ff7fe1e5298e))

## [1.3.0](https://github.com/coding327/webStorage/compare/v1.2.1...v1.3.0) (2023-06-23)


### Features

* 🚀 调整配置项，局部高于全局 ([87f2d80](https://github.com/coding327/webStorage/commit/87f2d806b867f39ee9e3da859532389ca4455055))
* 🚀 调整为1s ([27ad663](https://github.com/coding327/webStorage/commit/27ad6634aa51039cb700730ba90d024bacf28d26))

### [1.2.1](https://github.com/coding327/webStorage/compare/v1.2.0...v1.2.1) (2023-06-23)


### Bug Fixes

* 🧩 修复config不必传 ([562150f](https://github.com/coding327/webStorage/commit/562150f6cf6c02493e24c38248f34b8f8010b19c))

## [1.2.0](https://github.com/coding327/webStorage/compare/v1.1.3...v1.2.0) (2023-06-23)


### Features

* 🚀 吃吃吃 ([955bddb](https://github.com/coding327/webStorage/commit/955bddb3db269758d708c707b6ab0aa5e40a8e05))

### [1.1.3](https://github.com/coding327/webStorage/compare/v1.1.2...v1.1.3) (2023-06-23)

### [1.1.2](https://github.com/coding327/webStorage/compare/v1.1.1...v1.1.2) (2023-06-23)

### [1.1.1](https://github.com/coding327/webStorage/compare/v1.1.0...v1.1.1) (2023-06-23)

## 1.1.0 (2023-06-23)


### Features

* 🚀 init ([2c92bb5](https://github.com/coding327/webStorage/commit/2c92bb54e2955c80ef7eda243b0ca78aa2e7a35b))


src/crypto.ts
import * as CryptoJS from 'crypto-js';

const SECRET_KEY = CryptoJS.enc.Utf8.parse('3333e6e143439161'); //十六位十六进制数作为密钥
const SECRET_IV = CryptoJS.enc.Utf8.parse('e3bbe7e3ba84431a'); //十六位十六进制数作为密钥偏移量

const encrypt = (data: object | string): string => {
  //加密
  if (typeof data === 'object') {
    try {
      data = JSON.stringify(data);
    } catch (e) {
      throw new Error('encrypt error' + e);
    }
  }
  const dataHex = CryptoJS.enc.Utf8.parse(data);
  const encrypted = CryptoJS.AES.encrypt(dataHex, SECRET_KEY, {
    iv: SECRET_IV,
    mode: CryptoJS.mode.CBC,
    padding: CryptoJS.pad.Pkcs7,
  });
  return encrypted.ciphertext.toString();
};

const decrypt = (data: string) => {
  //解密
  const encryptedHexStr = CryptoJS.enc.Hex.parse(data);
  const str = CryptoJS.enc.Base64.stringify(encryptedHexStr);
  const decrypt = CryptoJS.AES.decrypt(str, SECRET_KEY, {
    iv: SECRET_IV,
    mode: CryptoJS.mode.CBC,
    padding: CryptoJS.pad.Pkcs7,
  });
  const decryptedStr = decrypt.toString(CryptoJS.enc.Utf8);
  return decryptedStr.toString();
};

export { encrypt, decrypt };


src/index-ot.ts
import { encrypt, decrypt } from './crypto';
import { globalConfig } from './types';

const config: globalConfig = {
  type: 'localStorage', //存储类型，localStorage | sessionStorage
  prefix: 'tmd-ui_0.0.1', //版本号
  expire: 24 * 60, //过期时间，默认为一天，单位为分钟
  isEncrypt: true, //支持加密、解密数据处理
};

const setStorage = (
  key: string,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  value: any,
  expire: number = 24 * 60,
): boolean => {
  //设定值
  if (value === '' || value === null || value === undefined) {
    //空值重置
    value = null;
  }
  if (isNaN(expire) || expire < 0) {
    //过期时间值合理性判断
    throw new Error('Expire must be a number');
  }
  const data = {
    value, //存储值
    time: Date.now(), //存储日期
    expire: Date.now() + 1000 * 60 * expire, //过期时间
  };
  //是否需要加密，判断装载加密数据或原数据
  window[config.type].setItem(
    autoAddPreFix(key),
    config.isEncrypt ? encrypt(JSON.stringify(data)) : JSON.stringify(data),
  );
  return true;
};

const getStorageFromKey = (key: string) => {
  //获取指定值
  if (config.prefix) {
    key = autoAddPreFix(key);
  }
  if (!window[config.type].getItem(key)) {
    //不存在判断
    return null;
  }

  const storageVal = config.isEncrypt
    ? JSON.parse(decrypt(window[config.type].getItem(key) as string))
    : JSON.parse(window[config.type].getItem(key) as string);
  const now = Date.now();
  if (now >= storageVal.expire) {
    //过期销毁
    removeStorageFromKey(key);
    return null;
    //不过期回值
  } else {
    return storageVal.value;
  }
};

const getAllStorage = () => {
  //获取所有值
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const storageList: any = {};
  const keys = Object.keys(window[config.type]);
  keys.forEach((key) => {
    const value = getStorageFromKey(autoRemovePreFix(key));
    if (value !== null) {
      //如果值没有过期，加入到列表中
      storageList[autoRemovePreFix(key)] = value;
    }
  });
  return storageList;
};

const getStorageLength = () => {
  //获取值列表长度
  return window[config.type].length;
};

const removeStorageFromKey = (key: string) => {
  //删除值
  if (config.prefix) {
    key = autoAddPreFix(key);
  }
  window[config.type].removeItem(key);
};

const clearStorage = () => {
  window[config.type].clear();
};

const autoAddPreFix = (key: string) => {
  //添加前缀，保持唯一性
  const prefix = config.prefix || '';
  return `${prefix}_${key}`;
};

const autoRemovePreFix = (key: string) => {
  //删除前缀，进行增删改查
  const lineIndex = config.prefix.length + 1;
  return key.substr(lineIndex);
};

export {
  setStorage,
  getStorageFromKey,
  getAllStorage,
  getStorageLength,
  removeStorageFromKey,
  clearStorage,
};


src/index.ts
import { encrypt, decrypt } from './crypto';
import { globalConfig } from './types';

class LocalCache {
  private config: globalConfig = {
    type: 'localStorage', //存储类型，localStorage | sessionStorage
    prefix: 'tmd-ui_0.0.1', //前缀
    expire: 24 * 60 * 60, //过期时间，默认为一天，单位为秒
    isEncrypt: true, //支持加密、解密数据处理
  };

  constructor(config?: globalConfig) {
    this.config = { ...this.config, ...config };
  }

  private autoAddPreFix(key: string): string {
    //添加前缀，保持唯一性
    const prefix = this.config.prefix || '';
    return `${prefix}_${key}`;
  }

  private autoRemovePreFix = (key: string) => {
    //删除前缀，进行增删改查
    const lineIndex = this.config.prefix.length + 1;
    return key.substring(lineIndex);
  };

  // 设置缓存
  // 有效期，默认为一天，单位为秒
  public setCache(
    key: string,
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    value: any,
    expire: number,
  ): boolean {
    //设定值
    if (value === '' || value === null || value === undefined) {
      //空值重置
      value = null;
    }
    expire = expire || this.config.expire || 24 * 60 * 60; //过期时间
    if (isNaN(expire) || expire < 0) {
      //过期时间值合理性判断
      throw new Error('Expire must be a number');
    }
    const data = {
      value, //存储值
      time: Date.now(), //存储日期
      expire: Date.now() + 1000 * expire, //过期时间
    };
    //是否需要加密，判断装载加密数据或原数据
    window[this.config.type].setItem(
      this.autoAddPreFix(key),
      this.config.isEncrypt
        ? encrypt(JSON.stringify(data))
        : JSON.stringify(data),
    );
    return true;
  }

  //获取指定值
  public getCache(key: string) {
    if (this.config.prefix) {
      key = this.autoAddPreFix(key);
    }
    if (!window[this.config.type].getItem(key)) {
      //不存在判断
      return null;
    }

    const storageVal = this.config.isEncrypt
      ? JSON.parse(decrypt(window[this.config.type].getItem(key) as string))
      : JSON.parse(window[this.config.type].getItem(key) as string);
    const now = Date.now();
    if (now >= storageVal.expire) {
      //过期销毁
      this.deleteCache(key);
      return null;
      //不过期回值
    } else {
      return storageVal.value;
    }
  }

  //获取全部值
  public getAllCache(): object {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const storageList: any = {};
    const keys = Object.keys(window[this.config.type]);
    keys.forEach((key) => {
      const value = this.getCache(this.autoRemovePreFix(key));
      if (value !== null) {
        //如果值没有过期，加入到列表中
        storageList[this.autoRemovePreFix(key)] = value;
      }
    });
    return storageList;
  }

  //获取缓存列表长度
  public getCacheLength(): number {
    return window[this.config.type].length;
  }

  //删除指定值
  public deleteCache(key: string): boolean {
    if (this.config.prefix) {
      key = this.autoAddPreFix(key);
    }
    window[this.config.type].removeItem(key);
    return true;
  }

  //清空所有缓存
  public clearCache(): boolean {
    window[this.config.type].clear();
    return true;
  }
}

export default LocalCache;
